src/coreclr/tools/aot/ILCompiler.ReadyToRun/Interop/IL/Marshaller.ReadyToRun.cs

```diff
@@ -25,12 +25,15 @@ protected static Marshaller CreateMarshaller(MarshallerKind kind)
                     return new BooleanMarshaller();
                 case MarshallerKind.AnsiString:
                     return new AnsiStringMarshaller();
+#if !READYTORUN
+                // UTF8 and Unicode Strings are not supported in R2R
                 case MarshallerKind.UTF8String:
                     return new UTF8StringMarshaller();
-                case MarshallerKind.SafeHandle:
-                    return new SafeHandleMarshaller();
                 case MarshallerKind.UnicodeString:
```

Comment:
Is this one not used in CoreLib? Looks like the common path of this marshaller is usable in crossgen - it's only corner cases (the string is passed by reference) that would hit non-existent helpers. We could also make the marshaller throw when it's in the corner case. (Just want to make sure we don't pregenerate fewer p/invokes and regress startup time.)

Reply:
Correct, its the non-pinned case here which is not handled correctly: https://github.com/dotnet/runtime/blob/fa779e8cb2b5868a0ac2fd4215f39ffb91f0dab0/src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs#L1494

Will check if throwing in that case excludes the method from precompilation.

Reply:
fixed to bring back UnicodeStringMarshaller